src/codegeneration/RewriteTailExpressionsTransformer.js

```diff
@@ -0,0 +1,163 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an 'AS IS' BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {ParseTreeTransformer} from './ParseTreeTransformer.js';
+import {
+  ArgumentList,
+  BinaryExpression,
+  CallExpression,
+  ConditionalExpression,
+  MemberExpression,
+  MemberLookupExpression
+} from '../syntax/trees/ParseTrees.js';
+import {
+  createArrayLiteralExpression,
+  createAssignmentExpression,
+  createCommaExpression,
+  createMemberExpression,
+  createIdentifierExpression as id,
+  createNullLiteral,
+  createParenExpression,
+} from './ParseTreeFactory.js';
+import {
+  COMMA_EXPRESSION,
+  MEMBER_EXPRESSION,
+  MEMBER_LOOKUP_EXPRESSION,
+  IDENTIFIER_EXPRESSION,
+  PAREN_EXPRESSION,
+  THIS_EXPRESSION
+} from '../syntax/trees/ParseTreeType.js';
+import {
+  AND,
+  OR
+} from '../syntax/TokenType.js';
+
+function createCall(tree, operand, thisArg) {
+  var argList = tree.args; // can be null
+  var argArray = argList ? argList.args : [];
+  argArray = argArray.map(arg => {
+    if (arg.type === COMMA_EXPRESSION) {
+      return createParenExpression(arg.type);
+    }
+    return arg;
+  });
+  return new CallExpression(tree.location,
+      createMemberExpression('$traceurRuntime', 'continuation'),
+      new ArgumentList(argList ? argList.location : null, [operand, thisArg,
+          createArrayLiteralExpression(argArray)]));
+}
+
+export class RewriteTailExpressionsTransformer extends ParseTreeTransformer {
+  // TODO(mnieper): template literals in tail position
+
+  constructor(bodyTransformer) {
+    this.bodyTransformer_ = bodyTransformer;
+  }
+
+  transformBinaryExpression(tree) {
+    var operator = tree.operator;
+    if (operator.type !== AND && operator.type !== OR) {
+      return tree;
+    }
+    var right = this.transformAny(tree.right);
+    if (right !== tree.right) {
+      return new BinaryExpression(tree.location, tree.left, operator, right);
+    }
+    return tree;
+  }
+
+  transformCallExpression(tree) {
+    var operand = tree.operand;
+    while (operand.type === PAREN_EXPRESSION) {
+      operand = operand.expression;
+    }
+    switch (operand.type) {
+      case IDENTIFIER_EXPRESSION:
+        return createCall(tree, operand, createNullLiteral());
+      case MEMBER_EXPRESSION:
+      case MEMBER_LOOKUP_EXPRESSION:
+        return this.transformMemberExpressionCall_(tree, operand)
+    }
+    return tree;
+  }
+
+  transformMemberExpressionCall_(tree, operand) {
+    var object = operand.operand;
+    var thisArg;
+    var assignment;
+    if (object.type === IDENTIFIER_EXPRESSION ||
+        object.type === THIS_EXPRESSION) {
+        // TODO(mnieper): More checks. Is there a better way?
```

Comment:
We have a lot of cases where we assume an IdentifierExpression has no side effect.